### PR TITLE
Docs: Application workload author's guide

### PR DESCRIPTION
**Pull request by sbose78**
1. Took out the annotation docs from https://github.com/redhat-developer/service-binding-operator/pull/628
2. Updated readme.
3. NEEDS REVIEW --> https://github.com/redhat-developer/service-binding-operator/pull/707/files#diff-ddcac56756c142138b380b68fe56e798

### CODE REVIEW

**Comment by pedjak** [comment]
/hold

#628 needs to be merged first

File: docs/application-author.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+  - group: "route.openshift.io"
+    version: v1
+    kind: Route
+    name: auth-service
+```
+
+The application workload expects binding metadata to be present on the Kubernetes Resources representing the backing service.
+
+As shown above, you may also directly use a `Configmap` or a `Secret` itsef as a service resource that would be used as a source of binding information.
```
Comment (Avni-Sharma) [inline docs/application-author.md:63]: nit - itself

Comment (Avni-Sharma) [inline docs/application-author.md:63]: nit- ConfigMap

File: docs/application-author.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+# Backing Service providing binding metadata
+
+If the backing service author has provided binding metadata in the corresponding CRD,
+then Service Binding acknowleges it and automatically creates a binding secret with
```
Comment (Avni-Sharma) [inline docs/application-author.md:68]: acknowledges

File: docs/application-author.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+the information relevant for binding.
+
+The backing service may provide binding information as
+* Metadata in the CRD as annotations
+* Metadata in the OLM bunde manifest file as Descriptiors
```
Comment (Avni-Sharma) [inline docs/application-author.md:73]: Descriptors

File: docs/application-author.md [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+* Secret or ConfigMap
+
+If the backing service provides binding metadata, you may use the resource as is
+to express an intent to bind your workload with one or more service resources, by creating a `ServiceBinding`.
+
+# Backing Service not providing binding metadata
+
+If the backing service hasn't provided any binding metadata, the application author may annotate the Kubernetes resource representating the backing service such that the managed binding secret generated has with the necessary binding information.
```
Comment (Avni-Sharma) [inline docs/application-author.md:81]: representing

Comment (Avni-Sharma) [reply docs/application-author.md:81]: has the necessary ...

File: docs/application-author.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+## Annotate Service Resources
+---
+
+The application author may consider specific elements of the backing service resource interesting for binding
+
+* A specific attribute in the `spec` section of the Kubernetes resource.
+* A specific attribute in the `status` section of the Kubernetes resource.
+* A specific attribute in the `data` section of the Kubernetes resource.
+* A specific attribute in a `Secret` referenced in the Kubernetes resource.
+* A specific attribute in a `Configmap` referenced in the Kubernetes resource.
```
Comment (Avni-Sharma) [inline docs/application-author.md:101]: ConfigMap

Comment (Avni-Sharma) [reply docs/application-author.md:101]: and other places where the same suggestion can be used

**Comment by sbose78** [comment]
Thank you @Avni-Sharma  , could you do another round of review, please?

**Comment by baijum** [comment]
/hold cancel

#628 merged

File: docs/annotations.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+### Data model : Building blocks for expressing binding information
+
+* `path`: A template representation of the path to the element in the Kubernetes resource. The value of `path` could be specified in either [JSONPath](https://kubernetes.io/docs/reference/kubectl/jsonpath/) or [GO templates](https://golang.org/pkg/text/template/)
+
+* `elementType`: Specifies if the value of the element referenced in `path` is of type `string` / `sliceOfStrings` / `sliceOfMaps`. Defaults to `string` if omitted.
+
+* `objectType`: Specifies if the value of the element indicated in `path` refers to a `ConfigMap`, `Secret` or a plain string in the current namespace!  Defaults to `Secret` if omitted and `elementType` is a non-`string`.
+
+* `bindAs`: Specifies if the element is to be bound as an environment variable or a volume mount using the keywords `envVar` and `volume`, respectively. Defaults to `envVar` if omitted.
```
Comment (baijum) [inline docs/annotations.md:25]:
```suggestion
* `bindAs`: Specifies if the element is bound as an environment variable or a volume mount using the keywords `envVar` and `volume`, respectively. Defaults to `envVar` if omitted.
```

File: docs/annotations.md [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+8. Extract a "slice of strings" from a Kubernetes resource and indicate the content in a specific index in the slice to be relevant for binding.
+
+
+### Data model : Building blocks for expressing binding information
+
+* `path`: A template representation of the path to the element in the Kubernetes resource. The value of `path` could be specified in either [JSONPath](https://kubernetes.io/docs/reference/kubectl/jsonpath/) or [GO templates](https://golang.org/pkg/text/template/)
+
+* `elementType`: Specifies if the value of the element referenced in `path` is of type `string` / `sliceOfStrings` / `sliceOfMaps`. Defaults to `string` if omitted.
+
+* `objectType`: Specifies if the value of the element indicated in `path` refers to a `ConfigMap`, `Secret` or a plain string in the current namespace!  Defaults to `Secret` if omitted and `elementType` is a non-`string`.
```
Comment (baijum) [inline docs/annotations.md:23]:
```suggestion
* `objectType`: Specifies if the value of the element indicated in `path` refers to a `ConfigMap`, `Secret`, or a plain string in the current namespace!  Defaults to `Secret` if omitted and `elementType` is a non-`string`.
```

Comment (baijum) [reply docs/annotations.md:23]: This is called "Oxford comma"

File: docs/annotations.md [hunk 8]
```diff
@@ -0,0 +1,286 @@
+## Binding Metadata in Annotations
+
+ During a binding operation, annotations from relevant Kubernetes resources are extracted to gather information about what is interesting for binding. This information is eventually used to bind the application with the backing service by populating the binding Secret.
+
+### Requirements for specifying binding information in a backing service CRD / Kubernetes resource
+
+1. Extract a string from the Kubernetes resource.
+2. Extract a string from the Kubernetes resource, and map it to custom name in the binding Secret.
```
Comment (baijum) [inline docs/annotations.md:8]:
```suggestion
2. Extract a string from the Kubernetes resource and map it to a custom name in the binding Secret.
```

File: docs/annotations.md [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+3. Extract an entire configmap/Secret from the Kubernetes resource.
+4. Extract a specific field from the configmap/Secret from the Kubernetes resource, and bind it as an environment variable.
+5. Extract a specific field from the configmap/Secret from the Kubernetes resource and and bind it as a volume mount.
```
Comment (baijum) [inline docs/annotations.md:11]:
```suggestion
5. Extract a specific field from the configmap/Secret from the Kubernetes resource and bind it as a volume mount.
```

File: docs/annotations.md [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+6. Extract a specific field from the configmap/Secret from the Kubernetes resource and map it to different name in the binding Secret.
```
Comment (baijum) [inline docs/annotations.md:12]:
```suggestion
6. Extract a specific field from the configmap/Secret from the Kubernetes resource and map it to a different name in the binding Secret.
```

File: docs/annotations.md [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+7. Extract a “slice of maps” from the Kubernetes resource and generate multiple fields in the binding Secret.
+8. Extract a "slice of strings" from a Kubernetes resource and indicate the content in a specific index in the slice to be relevant for binding.
```
Comment (baijum) [inline docs/annotations.md:14]:
```suggestion
8. Extract a "slice of strings" from a Kubernetes resource and indicate the content in a specific index in the slice relevant for binding.
```

File: docs/application-author.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+If the backing service author has provided binding metadata in the corresponding CRD,
+then Service Binding acknowledges it and automatically creates a binding secret with
+the information relevant for binding.
+
+The backing service may provide binding information as
+* Metadata in the CRD as annotations
+* Metadata in the OLM bunde manifest file as Descriptors
+* Secret or ConfigMap
+
+If the backing service provides binding metadata, you may use the resource as is
```
Comment (baijum) [inline docs/application-author.md:76]:
```suggestion
If the backing service provides binding metadata, you may use the resource
```

File: docs/annotations.md (hunk 6 above)
Comment (sbose78) [reply docs/annotations.md:25]: I did imply "is to be" implying "is supposed to be" or "is expected to be"

File: docs/application-author.md (hunk 1 above)
Comment (Avni-Sharma) [reply docs/application-author.md]: https://github.com/redhat-developer/service-binding-operator/pull/707/files#diff-60da86c468a457b0c38c7e96ee96ae29R9

and following sentences

File: docs/annotations.md [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+```
+
+
+
+### Scenarios
+
+
+1. #### Use everything from the Secret  “status.data.dbCredentials”
+
+    Requirement : *Extract an entire configmap/Secret from the Kubernetes resource*
```
Comment (Avni-Sharma) [inline docs/annotations.md:62]: ConfigMap

Comment (Avni-Sharma) [reply docs/annotations.md:62]: https://github.com/redhat-developer/service-binding-operator/pull/707/files#diff-60da86c468a457b0c38c7e96ee96ae29R85

Comment (Avni-Sharma) [reply docs/annotations.md:62]: https://github.com/redhat-developer/service-binding-operator/pull/707/files#diff-60da86c468a457b0c38c7e96ee96ae29R129

Comment (Avni-Sharma) [reply docs/annotations.md:62]: https://github.com/redhat-developer/service-binding-operator/pull/707/files#diff-60da86c468a457b0c38c7e96ee96ae29R152

**Comment by sbose78** [comment]
Thank you @baijum and @Avni-Sharma . I've addressed everything, I think.

**Pull request merged**